src/sage/schemes/hyperelliptic_curves/hyperelliptic_generic.py

```diff
@@ -248,29 +248,260 @@ def is_smooth(self):
+
+            sage: H.lift_x(3)
+            Traceback (most recent call last):
+            ...
+            ValueError: No point with x-coordinate 3 on Hyperelliptic Curve over Rational Field defined by y^2 = x^5 + x^3 + 1
+
+        The function also handles the case when `h(x)` is not zero::
+
+            sage: R.<x> = PolynomialRing(QQ)
+            sage: f = x^5 + x^3 + 1
+            sage: h = x + 1
+            sage: H = HyperellipticCurve(f, h)
+            sage: H.lift_x(1)
+            (1 : -3 : 1)
+
+        We can perform these operations over finite fields too::
+
+            sage: # needs sage.rings.finite_rings
+            sage: R.<x> = PolynomialRing(GF(163))
+            sage: f = x^7 + x + 1
+            sage: H = HyperellipticCurve(f)
+            sage: H.lift_x(13)
+            (13 : 41 : 1)
+
+        Including the case of characteristic two::
+
+            sage: # needs sage.rings.finite_rings
+            sage: F.<z4> = GF(2^4)
+            sage: R.<x> = PolynomialRing(F)
+            sage: f = x^7 + x^3 + 1
+            sage: h = x + 1
+            sage: H = HyperellipticCurve(f, h)
+            sage: H.lift_x(z4^3 + z4^2 + z4, all=True)
+            [(z4^3 + z4^2 + z4 : z4^2 + z4 + 1 : 1), (z4^3 + z4^2 + z4 : z4^3 : 1)]
+
+        TESTS::
+
+            sage: # needs sage.rings.finite_rings
+            sage: F1 = GF(11)
+            sage: F2 = GF(13)
+            sage: R.<x> = PolynomialRing(F1)
+            sage: f = x^7 + x^3 + 1
+            sage: H = HyperellipticCurve(f)
+            sage: H.lift_x(F2.random_element())
+            Traceback (most recent call last):
+            ...
+            ValueError: x must have a common parent with the base ring
+
+        Ensure that :issue:`37097` is fixed::
+
+            sage: # needs sage.rings.finite_rings
+            sage: F.<z4> = GF(2^4)
+            sage: R.<x> = PolynomialRing(F)
+            sage: f = x^7 + x^3 + 1
+            sage: h = x + 1
+            sage: H = HyperellipticCurve(f, h)
+            sage: H.lift_x(z4^3 + z4^2 + z4, all=True)
+            [(z4^3 + z4^2 + z4 : z4^2 + z4 + 1 : 1), (z4^3 + z4^2 + z4 : z4^3 : 1)]
+        """
+        from sage.structure.element import get_coercion_model
+        cm = get_coercion_model()
+
+        f, h = self.hyperelliptic_polynomials()
+        K = self.base_ring()
+
+        # Compute the common parent between the base ring of the curve and
+        # the parent of the input x-coordinate.
+        try:
+            L = cm.common_parent(x.parent(), K)
+            x = L(x)
+        except (TypeError, ValueError):
+            raise ValueError('x must have a common parent with the base ring')
+
+        # First we compute the y-coordinates the given x-coordinate
+        ys = []
+        one = L.one()
+
+        # When h is zero we find all y-coordinates with a single sqrt
+        if not h:
             y2 = f(x)
             if y2.is_square():
```

Comment:
I think the `.is_square()` check is unnecessary since `.sqrt(all=True)` will return the empty list when the element is not a square.

Reply:
This works for finite fields:

```py
sage: R.<x> = PolynomialRing(GF(13))
sage: f = x^5 + x^3 + 1
sage: y2 = f(3)
sage: y2.is_square()
False
sage: y2.sqrt(all=True)
[]
```

But for the rationals (or number fields) we get

```py
sage: R.<x> = PolynomialRing(QQ)
sage: f = x^5 + x^3 + 1
sage: y2 = f(3)
sage: y2.sqrt()
sqrt(271)
sage: y2.sqrt(all=True)
[sqrt(271), -sqrt(271)]
```

Reply:
(I'm not sure I like this behaviour, as I would expect QQ.sqrt() to return an element of QQ rathe than a `sage.symbolic.expression.Expression`, but that's a whole other thing to worry about)

Reply:
Passing `extend=False` should prevent the weird behaviour and throw a `ValueError` instead.

Reply:
...I guess the `ValueError` should actually be a `return []` in case `all=True` is also set...

Reply:
Inspired by:

```py
sage: K = QQ[I]
sage: K(3).sqrt(extend=False, all=True)
[]
sage: ZZ(3).sqrt(extend=False, all=True)
[]
sage: QQ(3).sqrt(extend=False, all=True)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [19], line 1
----> 1 QQ(Integer(3)).sqrt(extend=False, all=True)

File ~/sage/sage-10.2/src/sage/rings/rational.pyx:1971, in sage.rings.rational.Rational.sqrt()
   1969 if non_square:
   1970     if not extend:
-> 1971         raise ValueError("square root of %s not a rational number" % self)
   1972     from sage.misc.functional import _do_sqrt
   1973     return _do_sqrt(self, prec=prec, all=all)

ValueError: square root of 3 not a rational number
```

I have patched the sqrt for the rationals to behave like the integers and number fields so when extend is `False` and `all` is `True`, then `[]` is returned rather than an error.

I will rebuild and do tests etc then push this change.